### ISSUE TITLE
C++部署：instance_bank相关的疑问

### ISSUE THREAD

**Issue by yangwj2023** [ThomasVonWu/SparseEnd2End#14]
您好，我最近在基于010.sparse4d_end2end_io_bin_export.py流程移植到C++时，关于其中instance_bank有以下疑问：
 1. head网络中是否也包含一个instance_bank的功能？我看到在head推理中第一个refine算子执行后，会进入一个instance_bank.update()；
 2. instance_bank中的cache、update_track_id和decoder中基于类别置信度的策略均不一致，cache中参考了历史帧（排序后）的类别置信度，update_track_id中仅利用类别置信度选择前600，后300置为-1。另外，decoder时，参考了quality_score选择前300。这三者的差异原因是什么？
 
请大佬帮助答疑，我还是个小白，不知道是不是我对流程、或者接口功能理解错误，麻烦大佬帮忙解答，万分感谢！

**Comment by ThomasVonWu** [comment]
Hi, @yangwj2023 :

### Q1
instance_bank.update()实现和 head forward 确实有耦合:
```bash
if len(prediction) == self.num_single_frame_decoder:
    anchor = self.instance_bank.update(instance_feature, anchor, cls)
```
在部署过程需要将其实现移植入脚本[SparseEnd2End/deploy/export_head_onnx.py](https://github.com/ThomasVonWu/SparseEnd2End/blob/4ca34ce809291276241355a31629e65317024fee/deploy/export_head_onnx.py#L321):
```bash
if len(prediction) == self.num_single_frame_decoder:
    N = self.instance_bank.num_anchor - self.instance_bank.num_temp_instances
    cls = cls.max(dim=-1).values
    _, (selected_feature, selected_anchor) = topk(cls, N, instance_feature, anchor)
    selected_feature = torch.cat([temp_instance_feature, selected_feature], dim=1)
    selected_anchor = torch.cat([temp_anchor, selected_anchor], dim=1)
    instance_feature = torch.where(
        mask[:, None, None], selected_feature, instance_feature
    )
    anchor = torch.where(mask[:, None, None], selected_anchor, anchor)
    track_id = torch.where(
        mask[:, None],
        track_id,
        track_id.new_tensor(-1),
    )
```

## Q2
+ 1) `instance_bank.cache()` (***无论训练还是推理流都会走到***), 主要目的, 更新历史上一帧的instance相关三个维度的特征:
  + 首先, 缓存当前帧900个预设query的置信度预测值, 并基于长时序recurrent机制, 融合历史上一帧的900个预设query的置信度, 最后, 经过topK排序保留了融合后置信度前600的高分anchor, 并更新历史上一帧的instace(这里的instance包含三个维度: 实例锚框的语义特征self.cached_feature,  实例锚框的定位值self.cached_anchor和实例锚框的置信度:self.confidence). 注意, 这里关于实例instance相关的语义特征以及定位值是重新排序后的结果, 没有显式的特征融合.
+ 2) 缓存了历史上一帧的instance, 我们就可以利用模型做长时序的特征显式融合预测任务:
  + 2.1) head 每次一次forward过程中, 都会通过`instance_bank.get()`(***无论训练还是推理流都会走到***)拿到当前帧900个初始化instance和历史上一帧缓存的600个instance相关特征, 如下:
    + instance_feature
    + anchor
    + temp_instance_feature
    + temp_anchor
  + 2.2) 有了历史帧实例特征和当前帧的初始化的实例特征, 经过模型的前向的推理(cross-attention, refine), 我们最终可以拿到任务相关的当前帧预测实例instance的特征(instance_feature, instance_anchor, instance_confidence)
  + 2.3) 上一步中更新的当前帧`t`时刻实例特征, 将会送入instance_bank.cache(),  经过置信度融合重新排序, 我们继续更新历史上一帧`t-1`时刻的instance相关特征, 为下一帧的预测做准备.
这里我们分析下: head.refine具体做了啥, 他为什么需要和`instance_bank.update()`耦合:
  + 2.4) 既然我们定义它为refine, 那自然是为了精细的调整当前帧`t`时刻的instance相关特征, 作者这样设计: 首先将当前帧的instance按照置信度重新排序,选则前300个实例. 紧接着, 将历史上一帧的600个实例self.cached_feature `concat` 当前帧预测的300个实例instance feature, 将历史上一帧600个实例self.cached_anchor`concat` 当前帧预测的300个实例instance_anchor, 这样维度也能保持不变900个, 从而实现instance特征级精细化调整(不过concat这个操作看起来挺粗暴的):
    + selected_feature = torch.cat([self.cached_feature, selected_feature], dim=1)
    + selected_anchor = torch.cat([self.cached_anchor, selected_anchor], dim=1)
+ 3) 有了实例instance的instance_anchor, instance_feature, instance_classfication, instance_quality, 我们还缺个表示跟踪特性的变量:track_id.
  + 3.1) `instance_bank.update()` (***只有推理流都会走到***), 2.4中之所以构造600个实例self.cached_instance `concat` 当前帧预测的300个实例, 而不是构造 当前帧预测的300个实例instance `concat` 600个实例self.cached_instance, 是希望当前帧预测的实例的顺序是和历史上一帧的实例能够关联起来(简单点说希望他们对应位置是同一个障碍物), 有了这种隐式关联跟踪的关系, 我们在解码track_id的过程, 需要将历史上一帧的300个-1标志位的重新赋值新的trackid, 而前600保持当前的trackid.
  + 3.2) 由于实例轨迹会消失,也会新建, 下次解码trackid需要结合当前帧融合后的实例置信度`self.temp_confidence`更新历史上一帧的self.track_id, 参考 `instance_bank.update_track_id()` 代码逻辑: 作者保留600个位置给未来还存在轨迹的实例, 300个位置给新出现轨迹的实例, 前600个位置保留的trackid是通过topK排序保留置信度最大的最有可能继续存在的轨迹实例.

## Q3
+ decode 过程可以无视源码的实现, 后面我会更新的, 这里的置信度排序需要使用`self.temp_confidence`, 和quality_score&cls_scores都没有关系. 所以模型head-forward需要新增一个输出量`self.instance_bank.temp_confidence`.

希望以上回答对你有帮助!

**Comment by yangwj2023** [comment]
非常感谢大佬的讲解（详细、清晰，大赞），我又重新认识了这部分流程及功能，已经理解了（哈哈哈......）。
另外，期待大佬更新C++部署代码，供我们这些小白们学习、参考（我个人能力有限，python翻译过来，运行结果差别较大（结果不对，且运行时间较长，总耗时将近100ms）。

**Comment by ThomasVonWu** [comment]
C++ 部署之前用过3090做过一轮测试, 效果还可以, 不过耗时确实比较长, Backbone需要做量化, head还需要加速, 精度差异到没有出现太大差异性. 我尽快将C++ Baseline 代码完善推上来.

**Comment by yangwj2023** [comment]
好的，非常感谢大佬，您以上讲解对我有很大帮助（大赞哇）。
